data/src/main/java/com/linkedin/data/schema/annotation/DataSchemaRichContextTraverser.java

```diff
@@ -110,15 +105,16 @@ private void doRecursiveTraversal(TraverserContextImpl context)
       _seenAncestorsDataSchema.put(schema, Boolean.TRUE);
 
       // Pass new context in every recursion
-      TraverserContextImpl nextContext = null;
-
+      TraverserContextImpl nextContext;
       switch (schema.getType())
       {
         case TYPEREF:
           TyperefDataSchema typerefDataSchema = (TyperefDataSchema) schema;
-
-          nextContext = context.getNextContext(DataSchemaConstants.REF_KEY, null, typerefDataSchema.getRef(),
-                                               CurrentSchemaEntryMode.TYPEREF_REF);
+          String nextSchemaPathComponent =
+              _schemaVisitor.shouldIncludeTyperefsInPathSpec() ? DataSchemaConstants.TYPEREF_REF : null;
```

Comment:
I am afraid adding this `shouldIncludeTyperefsInPathSpec` to `SchdemaVisitor` and `DataSchemaRichContextTraverser` might make things a bit confusing because as we defined in `TraverserContext`, the `SchemaPathSpec` should not include typeref in pathSpec.

I know previously `AnnotationCompatibilityChecker` did a trick in its implementation to override that, but I doubt if we should add it as a method in `schemaVisitor` ?

Reply:
I thought of other ways to resolve it but couldn't find anything non-invasive. Any suggestions?

Reply:
As discussed offline added a doc to explain this better.